### PR TITLE
Added support for directory and delegation SAS

### PR DESCRIPTION
**Pull request by gapra-msft**
This PR resolves https://github.com/Azure/azure-sdk-for-java/issues/10378

### CODE REVIEW

File: sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/implementation/Constants.java [hunk 1]
```diff
@@ -347,6 +347,26 @@ private HeaderConstants() {
          */
         public static final String SAS_SIGNED_KEY_VERSION = "skv";
 
+        /**
+         * The SAS AAD object id parameter for user delegation SAS.
+         */
+        public static final String SAS_AAD_OBJECT_ID = "saoid";
```
Comment (gapra-msft) [inline sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/implementation/Constants.java:353]: rename to SAS_AUTHORIZED_AAD_OBJECT_ID

File: sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/sas/CommonSasQueryParameters.java [hunk 2]
```diff
@@ -62,6 +63,14 @@
 
     private final String contentType;
 
+    private final Integer directoryDepth;
+
+    private final String aadObjectId;
```
Comment (gapra-msft) [inline sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/sas/CommonSasQueryParameters.java:68]: rename to authorized

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/implementation/util/BlobSasImplUtil.java [hunk 3]
```diff
@@ -110,7 +110,7 @@ public BlobSasImplUtil(BlobServiceSasSignatureValues sasValues, String container
             throw logger.logExceptionAsError(
                 new IllegalArgumentException("'snapshot' and 'versionId' cannot be used at the same time."));
         }
-        this.version = sasValues.getVersion();
+        this.version = null;
```
Comment (rickle-msft) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/implementation/util/BlobSasImplUtil.java:113]: Can you comment these lines to explain we set it to null to force the latest version?

File: sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/SasClientTests.groovy [hunk 4]
```diff
@@ -209,7 +209,7 @@ class SasClientTests extends APISpec {
         validateSasProperties(properties)
     }
 
-    @Ignore("User Delegation SAS string to sign has changed. Reenable when working on feature.")
+//    @Ignore("User Delegation SAS string to sign has changed. Reenable when working on feature.")
```
Comment (rickle-msft) [inline sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/SasClientTests.groovy:212]: Delete

File: sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/implementation/util/DataLakeSasImplUtil.java [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+            tryAppendQueryParameter(sb, Constants.UrlConstants.SAS_SIGNED_KEY_START,
+                formatQueryParameterDate(userDelegationKey.getSignedStart()));
+            tryAppendQueryParameter(sb, Constants.UrlConstants.SAS_SIGNED_KEY_EXPIRY,
+                formatQueryParameterDate(userDelegationKey.getSignedExpiry()));
+            tryAppendQueryParameter(sb, Constants.UrlConstants.SAS_SIGNED_KEY_SERVICE,
+                userDelegationKey.getSignedService());
+            tryAppendQueryParameter(sb, Constants.UrlConstants.SAS_SIGNED_KEY_VERSION,
+                userDelegationKey.getSignedVersion());
+
+            /* Only parameters relevant for user delegation SAS. */
```
Comment (rickle-msft) [inline sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/implementation/util/DataLakeSasImplUtil.java:204]: Is this comment not true for the whole if block?

**Review by rickle-msft — APPROVED** [review]

File: sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/SasClientTests.groovy (hunk 4 above)
Comment (gapra-msft) [reply sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/SasClientTests.groovy:212]: deleted

File: sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/implementation/util/DataLakeSasImplUtil.java (hunk 5 above)
Comment (gapra-msft) [reply sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/implementation/util/DataLakeSasImplUtil.java:204]: Yeah I wanted to more be like these are the only params relevant to user delegation sas but not actually tied to the user delegation key

**Review by amnguye — APPROVED** [review]

**Pull request merged**